File: Source/Tortuga/Graphics/Vulkan/DescriptorLayout.cpp
#include "./DescriptorLayout.hpp"

/*
struct Descriptor { type specific data }

struct DescriptorBinding {   
  int binding;
  DescriptorType type;
  Descriptor descriptors[]
};

struct DescriptorSet {
    DescriptorBinding bindings[];
};

struct PipelineLayout {
  DescriptorSet sets[]
}
*/

namespace Tortuga
{
namespace Graphics
{
namespace Vulkan
{
namespace DescriptorLayout
{
DescriptorLayout Create(Device::Device device, uint32_t bindingsAmount, VkShaderStageFlags shaderStage, VkDescriptorType type)
{
  DescriptorLayout data = {};
  data.Device = device.Device;
  data.BindingsAmount = bindingsAmount;
  data.Type = type;

  std::vector<VkDescriptorSetLayoutBinding> pBindings(bindingsAmount);
  for (uint32_t i = 0; i < bindingsAmount; i++)
  {
    //bindings
    pBindings[i].binding = i;
    pBindings[i].descriptorType = type;
    pBindings[i].descriptorCount = 1;
    pBindings[i].stageFlags = shaderStage;
    pBindings[i].pImmutableSamplers = VK_NULL_HANDLE;
  }
  VkDescriptorSetLayoutCreateInfo createInfo = {};
  {
    createInfo.sType = VK_STRUCTURE_TYPE_DESCRIPTOR_SET_LAYOUT_CREATE_INFO;
    createInfo.bindingCount = pBindings.size();
    createInfo.pBindings = pBindings.data();
  }
  ErrorCheck::Callback(vkCreateDescriptorSetLayout(device.Device, &createInfo, nullptr, &data.Layouts));
  return data;
}
void Destroy(DescriptorLayout data)
{
  if (data.Layouts == VK_NULL_HANDLE)
    return;

  vkDestroyDescriptorSetLayout(data.Device, data.Layouts, nullptr);
}
} // namespace DescriptorLayout
} // namespace Vulkan
} // namespace Graphics
} // namespace Tortuga

File: Tortuga.Graphics/API/DescriptorLayout.cs
#pragma warning disable CS1591
using System;
using System.Collections.Generic;
using Tortuga.Utils;
using Vulkan;

/*
-- GPU DATA STRUCTURE --
struct Descriptor { type specific data }
struct DescriptorBinding {   
  int binding;
  DescriptorType type;
  Descriptor descriptors[]
};

struct DescriptorSet {
    DescriptorBinding bindings[];
};

struct PipelineLayout {
  DescriptorSet sets[]
}
*/

namespace Tortuga.Graphics.API
{
    public struct DescriptorBindingInfo
    {
        public uint Index;
        public VkDescriptorType DescriptorType;
        public uint DescriptorCounts;
        public VkShaderStageFlags ShaderStageFlags;

        public DescriptorBindingInfo(
            uint index,
            VkDescriptorType descriptorType,
            uint descriptorCounts,
            VkShaderStageFlags shaderStageFlags
        )
        {
            Index = index;
            DescriptorType = descriptorType;
            DescriptorCounts = descriptorCounts;
            ShaderStageFlags = shaderStageFlags;
        }
    }

    public class DescriptorLayout
    {
        public Device Device => _device;
        public VkDescriptorSetLayout Handle => _handle;
        public List<DescriptorBindingInfo> Bindings => _bindings;

        private Device _device;
        private VkDescriptorSetLayout _handle;
        private List<DescriptorBindingInfo> _bindings;

        public unsafe DescriptorLayout(Device device, List<DescriptorBindingInfo> bindings)
        {
            _device = device;
            _bindings = bindings;

            var vulkanBindings = new NativeList<VkDescriptorSetLayoutBinding>();
            foreach (var binding in bindings)
            {
                vulkanBindings.Add(new VkDescriptorSetLayoutBinding
                {
                    binding = binding.Index,
                    descriptorType = binding.DescriptorType,
                    descriptorCount = binding.DescriptorCounts,
                    stageFlags = binding.ShaderStageFlags
                });
            }

            var createInfo = new VkDescriptorSetLayoutCreateInfo
            {
                sType = VkStructureType.DescriptorSetLayoutCreateInfo,
                bindingCount = vulkanBindings.Count,
                pBindings = (VkDescriptorSetLayoutBinding*)vulkanBindings.Data.ToPointer()
            };

            VkDescriptorSetLayout layout;
            if (VulkanNative.vkCreateDescriptorSetLayout(
                device.Handle,
                &createInfo,
                null,
                &layout
            ) != VkResult.Success)
                throw new Exception("failed to create descriptor set layout");
            _handle = layout;
        }
        unsafe ~DescriptorLayout()
        {
            if (_handle != VkDescriptorSetLayout.Null)
            {
                VulkanNative.vkDestroyDescriptorSetLayout(
                    _device.Handle,
                    _handle,
                    null
                );
                _handle = VkDescriptorSetLayout.Null;
            }
        }
    }
}